From a5c4a37b19bd1831b7d28b97d51ef4ca83aaf960 Mon Sep 17 00:00:00 2001
From: Vincent Ahlrichs <vincent.ahlrichs@aisec.fraunhofer.de>
Date: Wed, 20 Jul 2022 14:03:34 +0200
Subject: [PATCH 2/2] riscv/vdso-offsets: Add header for cheri offsets

The offsets provided in the new header are generated based on the
cheri vDSO. As the offsets differ from the non-cheri vDSO, the header
corresponding the usage of CHERI must be included, to point to the
correct vDSO function symbol.
---
 arch/riscv/Makefile                   | 7 +++++++
 arch/riscv/include/asm/vdso.h         | 7 ++++++-
 arch/riscv/kernel/vdso-cheri/Makefile | 2 +-
 3 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/arch/riscv/Makefile b/arch/riscv/Makefile
index 2b4270153bb3..86d964f8d426 100644
--- a/arch/riscv/Makefile
+++ b/arch/riscv/Makefile
@@ -124,9 +124,16 @@ vdso_install:
 
 ifeq ($(KBUILD_EXTMOD),)
 ifeq ($(CONFIG_MMU),y)
+ifeq ($(CONFIG_RISCV_CHERI), y)
 prepare: vdso_prepare
 vdso_prepare: prepare0
 	$(Q)$(MAKE) $(build)=arch/riscv/kernel/vdso include/generated/vdso-offsets.h
+	$(Q)$(MAKE) $(build)=arch/riscv/kernel/vdso-cheri include/generated/vdso-cheri-offsets.h
+else
+prepare: vdso_prepare
+vdso_prepare: prepare0
+	$(Q)$(MAKE) $(build)=arch/riscv/kernel/vdso include/generated/vdso-offsets.h
+endif
 endif
 endif
 
diff --git a/arch/riscv/include/asm/vdso.h b/arch/riscv/include/asm/vdso.h
index 208e31bc5d1c..83688ee14aca 100644
--- a/arch/riscv/include/asm/vdso.h
+++ b/arch/riscv/include/asm/vdso.h
@@ -25,11 +25,16 @@
 #define __VVAR_PAGES    1
 
 #ifndef __ASSEMBLY__
+#ifndef __CHERI_PURE_CAPABILITY__
 #include <generated/vdso-offsets.h>
-
 #define VDSO_SYMBOL(base, name)							\
 	(void __user *)((unsigned long)(base) + __vdso_##name##_offset)
+#else
+#include <generated/vdso-cheri-offsets.h>
 
+#define VDSO_SYMBOL(base, name)							\
+	(void __user *)((unsigned long)(base) + __vdso_##name##_offset)
+#endif
 #endif /* CONFIG_MMU */
 
 #endif /* !__ASSEMBLY__ */
diff --git a/arch/riscv/kernel/vdso-cheri/Makefile b/arch/riscv/kernel/vdso-cheri/Makefile
index 609bf594db11..f8c98365b290 100644
--- a/arch/riscv/kernel/vdso-cheri/Makefile
+++ b/arch/riscv/kernel/vdso-cheri/Makefile
@@ -66,7 +66,7 @@ gen-vdsosym := $(srctree)/$(src)/gen_vdso_offsets.sh
 quiet_cmd_vdsosym = VDSOSYM $@
 	cmd_vdsosym = $(NM) $< | $(gen-vdsosym) | LC_ALL=C sort > $@
 
-include/generated/vdso-offsets.h: $(obj)/vdso.so.dbg FORCE
+include/generated/vdso-cheri-offsets.h: $(obj)/vdso.so.dbg FORCE
 	$(call if_changed,vdsosym)
 
 # actual build commands
-- 
2.37.3

